### PR TITLE
refactor(binarypb): handlers are context managers

### PR DESCRIPTION
**Pull request by cristianmtr**
To be merged only **after** https://github.com/jina-ai/jina/pull/2102

### CODE REVIEW

File: jina/executors/indexers/keyvalue.py [hunk 1]
```diff
@@ -2,9 +2,11 @@
 __license__ = "Apache-2.0"
 
 import mmap
+import os
 from typing import Iterable, Optional
 
 import numpy as np
+from tensorflow.python.autograph.pyct.anno import keys
```
Comment (JoanFM) [inline jina/executors/indexers/keyvalue.py:9]: tensorflow is not a basic dependency of us

Comment (cristianmtr) [reply jina/executors/indexers/keyvalue.py:9]: Don't know how that got in there. Some PyCharm auto-import maybe

File: jina/executors/indexers/keyvalue.py [hunk 2]
```diff
@@ -13,24 +14,49 @@
...
+
+        def _close(self):
+            # TODO do we need to make sure
             self.body.close()
             self.header.close()
 
-        def flush(self):
+        def close(self):
+            """Close the file."""
+            if self.__dict__.get('body') and not self.body.closed:
```
Comment (JoanFM) [inline jina/executors/indexers/keyvalue.py:52]: more pythonic way is to do `getattr(self, body, None)`

File: jina/executors/indexers/keyvalue.py [hunk 3]
```diff
@@ -13,24 +14,49 @@
...
+                self.body.flush()
+                self.body.close()
+            if self.__dict__.get('header') and not self.header.closed:
```
Comment (JoanFM) [inline jina/executors/indexers/keyvalue.py:55]: same here

**Comment by cristianmtr** [comment]
Closing this, as it is not possible to have context handlers Read/Write Handlers in BinaryPb.

Why?

We are relying on `write_handler`, `query_handler` fields to be callable in the `close` of the `BaseIndexer`, with a `.close()` call of their own. This would not be possible in the ctx mger approach.

Also, using a ctx mger is less efficient for `query`, as we would need to always rebuild the `query_handler` for every `query` call.

**Comment by JoanFM** [comment]
> Closing this, as it is not possible to have context handlers Read/Write Handlers in BinaryPb.
> 
> Why?
> 
> We are relying on `write_handler`, `query_handler` fields to be callable in the `close` of the `BaseIndexer`, with a `.close()` call of their own. This would not be possible in the ctx mger approach.
> 
> Also, using a ctx mger is less efficient for `query`, as we would need to always rebuild the `query_handler` for every `query` call.

Are we sure this is impossible? Isn't maybe the problem that BaseIndexer is assuming something that is not true?

**Comment by cristianmtr** [comment]
> > Closing this, as it is not possible to have context handlers Read/Write Handlers in BinaryPb.
> > Why?
> > We are relying on `write_handler`, `query_handler` fields to be callable in the `close` of the `BaseIndexer`, with a `.close()` call of their own. This would not be possible in the ctx mger approach.
> > Also, using a ctx mger is less efficient for `query`, as we would need to always rebuild the `query_handler` for every `query` call.
> 
> Are we sure this is impossible? Isn't maybe the problem that BaseIndexer is assuming something that is not true?

Could be. If we do go down that path, it would require a rewrite of `BaseIdxer` though. Maybe that's a bit outside our current scope

Also, the statement about speed/performance would still hold. We would be opening the ReadHandler on every update, for each document

**Comment by JoanFM** [comment]
> > > Closing this, as it is not possible to have context handlers Read/Write Handlers in BinaryPb.
> > > Why?
> > > We are relying on `write_handler`, `query_handler` fields to be callable in the `close` of the `BaseIndexer`, with a `.close()` call of their own. This would not be possible in the ctx mger approach.
> > > Also, using a ctx mger is less efficient for `query`, as we would need to always rebuild the `query_handler` for every `query` call.
> > 
> > 
> > Are we sure this is impossible? Isn't maybe the problem that BaseIndexer is assuming something that is not true?
> 
> Could be. If we do go down that path, it would require a rewrite of `BaseIdxer` though. Maybe that's a bit outside our current scope
> 
> Also, the statement about speed/performance would still hold. We would be opening the ReadHandler on every update, for each document

there may be a way to keep it as context manager and load it only when needed, but maybe it needs further refactoring

**Comment by JoanFM** [comment]
At least we need to surround the usage with try catches

**Pull request closed without merging**